File: pkg/compiler/templates/flow8.go
package main

import "fmt"

// nested calls: can panic transfer
// from lower in the stack to higher up?

var global int = 8

func deeper(a int) {
	global += (a + 3)
	panic("panic-in-deeper")
}

func intermed(a int) {
	deeper(a)
}

func main() {
	a := 0
	b := 0

	f := func() (ret0, ret1 int) {

		defer func(a int) {
			fmt.Printf("first defer running, a=%v, b=%v, ret0=%v, ret1=%v, global=%v\n",
				a, b, ret0, ret1, global)
			b = b + 3
			ret0 = (ret0+1)*3 + global
			fmt.Printf("debug, after ret0 = (ret0+1) * 3 + global: ret0=%v", ret0)
			ret1 = ret1 + 1
			recov := recover()
			fmt.Printf("defer 1 recovered '%v'\n", recov)
			if recov != nil {
				fmt.Printf("recov was not nil, ret0=%v  and ret1=%v", ret0, ret1)
				ret1 = ret1 + 9 + global
				ret0 = ret0 + 19 + global
			}
			fmt.Printf("at end of 1st defer, ret0='%v', ret1='%v'\n", ret0, ret1)
		}(a)

		intermed(a)
		return
	}
	f1, f2 := f()
	fmt.Printf("f1 = %v, f2 = %v\n", f1, f2)
	/*
	go run flow8.go
	first defer running, a=0, b=0, ret0=0, ret1=0, global=11
	debug, after ret0 = (ret0+1) * 3 + global: ret0=14defer 1 recovered 'panic-in-deeper'
	recov was not nil, ret0=14  and ret1=1at end of 1st defer, ret0='44', ret1='21'
	f1 = 44, f2 = 21
	*/
}

File: pkg/compiler/_templates/flow8.lua
dofile 'deferinit.lua'


a = 0;
b = 0;

-- nested calls: can panic transfer
-- from lower in the stack to higher up?

global = 0


function deeper(...)
   orig = {...}
   
   local __defers={}
   local __zeroret = {}
   local __namedNames = {}

  local __actual=function(a)
     global = global + (a + 3)
     panic("panic-in-deeper")
  end -- end of __actual
  return __actuallyCall("deeper", __actual, __namedNames, __zeroret, __defers)
end


function intermed(a)
   deeper(a)
end



function f(...)
   orig = {...}
   local __defers={}
   local __zeroret = {0,0}
   local __namedNames = {"ret0", "ret1"}   
   
   local __actual=function()

      local __defer_func = function(a)
         
         -- capture any arguments at defer call point
         local a = a
         __defers[1+#__defers] = function() 
           print("first defer running, a=", a, " b=",b, " ret0=", ret0, " ret1=", ret1)
            b = b + 3
            ret0 = (ret0+1) * 3 + global
            ret1 = ret1 + 1
            recov = recover()
            print("defer 1 recovered ", recov)
            if recov ~= nil then
               ret1 = ret1 + 9 + global
               ret0 = ret0 + 19 + global
            end
        end
     end
     __defer_func(a)

     deeper(a)
     return

   end -- end of __actual
   return __actuallyCall("f", __actual, __namedNames, __zeroret, __defers)
end

f1, f2 = f()
print("f1 = ",f1, " f2=", f2)

--[[
dofile 'flow8.lua'
deeper: panicHandler running with err =	flow8.lua:30: panic-in-deeper	 and #defer = 	0
deeper: checking for panic still un-caught...	flow8.lua:30: panic-in-deeper
deeper: un recovered error still exists, rethrowing 	flow8.lua:30: panic-in-deeper
panicHandler running with err =	flow8.lua:89: flow8.lua:30: panic-in-deeper	 and #defer = 	1
first defer running, a=	0	 b=	0	 ret0=	0	 ret1=	0
defer 1 recovered 	flow8.lua:89: flow8.lua:30: panic-in-deeper
panic path defer call result: i=	1	  v=	true
checking for panic still un-caught...	nil
at f return point, ret0 = 	28	 ret1=	13
f1 = 	28	 f2=	13

--]]
